### PR TITLE
Slack notifier

### PR DESCRIPTION
**Pull request by ssd04**
Add slack notifier. Send events to slack via webhook integration.

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -40,6 +42,12 @@ type Email struct {
 	To            []string
 }
 
+// Slack holds the configuration for slack notifier
+type Slack struct {
+	Enabled bool
+	URL     string
```
Comment (bogdan-rosianu) [inline config/config.go:48]: the API key will be contained in the URL, right?

File: notifiers/slack/notifier.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+	return nil
+}
+
+// PushMessage will push the notification
+func (sn *slackNotifier) PushMessage(msg data.NotificationMessage) error {
+	return sn.push(msg)
+}
+
+func (sn *slackNotifier) push(msg data.NotificationMessage) error {
+	data := payload{
```
Comment (miiu96) [inline notifiers/slack/notifier.go:51]: Variable 'data' collides with imported package name

File: notifiers/slack/notifier.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+	if err != nil {
+		return err
+	}
+	req.Header.Set("Content-Type", "application/json")
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
```
Comment (miiu96) [inline notifiers/slack/notifier.go:70]: Unhandled error

File: config/config.go (hunk 1 above)
Comment (ssd04) [reply config/config.go:48]: yes :+1: in config there is also an example with the url format

File: notifiers/slack/notifier.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+	data := payload{
+		Text: msg.Message,
+	}
+	payloadBytes, err := json.Marshal(data)
+	if err != nil {
+		return err
+	}
+	body := bytes.NewReader(payloadBytes)
+
+	req, err := http.NewRequest("POST", sn.config.URL, body)
```
[the remark this answers is not in the document]
Comment (ssd04) [reply notifiers/slack/notifier.go]: done

File: notifiers/slack/notifier.go (hunk 2 above)
Comment (ssd04) [reply notifiers/slack/notifier.go]: renamed variable

File: notifiers/slack/notifier.go (hunk 3 above)
Comment (ssd04) [reply notifiers/slack/notifier.go]: done

File: notifiers/slack/notifier.go [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+type ArgsSlackNotifier struct {
+	Config *config.Slack
+}
+
+type slackNotifier struct {
+	config *config.Slack
+}
+
+// NewSlackNotifier will create a new email notifier instance
+func NewSlackNotifier(args ArgsSlackNotifier) (*slackNotifier, error) {
```
Comment (iulianpascalau) [inline notifiers/slack/notifier.go:29]: some unit tests would have been great

**Review by iulianpascalau — CHANGES_REQUESTED** [review]

File: monitoring/monitoringRunner.go [hunk 6]
```diff
@@ -52,6 +53,15 @@ func (mr *monitoringRunner) Start() error {
 
 	notifyProcessor := process.NewNotifyProcessor()
 
+	if mr.config.Notifiers.Slack.Enabled {
+		argsSlackNotifier := slack.ArgsSlackNotifier{Config: mr.config.Notifiers.Slack}
```
[the remark this answers is not in the document]
Comment (ssd04) [reply monitoring/monitoringRunner.go:57]: :+1:

File: notifiers/slack/notifier.go [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+// PushMessage will push the notification
+func (sn *slackNotifier) PushMessage(msg data.NotificationMessage) error {
+	return sn.push(msg)
+}
+
+func (sn *slackNotifier) push(msg data.NotificationMessage) error {
+	msgPayload := payload{
+		Text: msg.Message,
+	}
+	payloadBytes, err := json.Marshal(msgPayload)
```
[the remark this answers is not in the document]
Comment (ssd04) [reply notifiers/slack/notifier.go]: moved this code to http client wrapper

**Review by bogdan-rosianu — DISMISSED** [review]

File: clients/httpClient.go [hunk 8]
```diff
@@ -73,6 +75,44 @@ func (hcw *httpClientWrapper) CallGetRestEndPoint(
 	return ioutil.ReadAll(resp.Body)
 }
 
+// CallPostRestEndPoint calls an external end point
+func (hcw *httpClientWrapper) CallPostRestEndPoint(
+	address string,
```
Comment (iulianpascalau) [inline clients/httpClient.go:80]: could have been only one string param here as we do a + anyway (valid also for the get method).
Can be left as it is for now

File: notifiers/slack/notifier.go [hunk 9]
```diff
@@ -18,66 +15,46 @@ type payload struct {
...
 
 type slackNotifier struct {
-	config *config.Slack
+	url        string
+	httpClient notifiers.HTTPClient
 }
 
 // NewSlackNotifier will create a new email notifier instance
-func NewSlackNotifier(args ArgsSlackNotifier) (*slackNotifier, error) {
+func NewSlackNotifier(args *ArgsSlackNotifier) (*slackNotifier, error) {
```
Comment (iulianpascalau) [inline notifiers/slack/notifier.go:28]: could have kept the argument DTO passed as value (same reason: a client struct can not alter the same argument used - maybe - on another client struct)

**Review by iulianpascalau — APPROVED** [review]

File: clients/httpClient.go (hunk 8 above)
Comment (ssd04) [reply clients/httpClient.go:82]: added a TODO for this

File: notifiers/slack/notifier.go (hunk 9 above)
Comment (ssd04) [reply notifiers/slack/notifier.go]: changed to pass the argument as value

**Review by bogdan-rosianu — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Pull request merged**